### ISSUE TITLE
AST Serialization for improving Build performance

### ISSUE THREAD

**Issue by CircArgs** [DataJunction/dj#688]
In order to reuse all the work done by compilation at node creating time and cache it, we should have a means to serialize/deserialize ASTs and store them with a `NodeRevision`.

In the interest of keeping the serialization to the bare minimum, I think the best methodology would be a dedicated serialization method for any `ast.Node` that has `is_compiled` which looks at a distinct set of attributes. This would ensure that we only serialize what is needed.

**Comment by shangyian** [comment]
@CircArgs agreed, thanks for writing up this issue! Have you already started playing around with this or shall I poke at it?

I had a closer look at the things we need to track for the subclasses of `ast.Node`. We should be able to serialize into a nested JSON structure with these base objects:
```
Alias:
  * child: Node
  * alias
  * name
  * as_

TableExpression:
  * in_from_or_lateral
  * columns

Table:
  * in_from_or_lateral
  * columns
  * dj_node [ I don't think we need this as all columns + types should've been extracted already, but we can keep the node name ]

Column:
  * alias
  * name
  * expression
  * table [we can store the table, but I'm not sure it's needed if we already assign all the columns types]
  * type
  * is_api_column

Expression:
  * columns: List[Column]
  * parenthesized
  * type
  * is_aggregation

Name:
  * name: str
  * quote_style
  * namespace
 
Named:
  * name: Name
  * columns [ possibly unneeded ]
  * names [ possibly unneeded as can be derived from name ]
  * namespace [ possibly unneeded as can be derived from name ]

JoinCriteria:
  * on: Expression
  * using: List[Named]

Join:
  * join_type: str
  * criteria: JoinCriteria
  * lateral: bool
  * natural: bool

LateralView:
  * outer
  * func

Relation:
  * primary
  * extensions: List[Join]

From:
  * relations: List[Relation]

SetOp:
  * kind
  * right

SortItem:
  * expression
  * asc
  * nulls

Organization:
  * order
  * sort
```

There's probably more that I've missed, but those should all be pretty straightforward renditions of the member variables of each class.

**Comment by CircArgs** [comment]
@shangyian I haven't started yet, but I think it will be pretty straightforward. I figured, like my short write-up suggested, that `Node` might handle a pretty general case and only the necessary subclasses would need special functionality like your summary clarifies - we only really care about facilitating functionality after compile which I hope means we only need to maintain attributes that signal a node is "compiled". There is some general subtleties to possibly handle, for example, Nodes also track the attribute name they are the child for, which is used in some methods like when nodes belong to a list on their parent, but that could be resolved when deserializing.
